docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,264 @@
+of interest.
+
+# The type systems
+
+The key difference between these 2 tools is the "type system". While both are exposing Rust
+code (which obviously comes with its own type system), the foreign bindings need to know
+lots of details about all the types expressed by the tool.
+
+For the sake of this document, we will use the term "type universe" to define the set of
+all types known by each of the tools. Both of these tools build their own "type universe" then
+use that to generate both Rust code and foreign bindings.
+
+## UniFFI's type universe
+
+UniFFI's model is to parse an external ffi description from a `.udl` file which describes the
+entire "type universe". This type universe is then used to generate both the Rust scaffolding
+(on disk as a `.rs` file) and the foreign bindings.
+
+**What's good about this** is that the entire type system is known when generating both the rust code
+and the foreign binding, and is known without parsing any Rust code. This is important because
+things like field names and types in structs must be known on both sides of the FFI.
+
+**What's bad about this** is that the external UDL is very ugly and redundant in terms of the
+implemented rust API.
+
+## Diplomat's type universe
+
+Diplomat defines its "type universe" (ie, the external ffi) using macros.
+
+**What's good about this** is that an "ffi module" (and there may be many) defines the canonical API
+and it is defined in terms of Rust types - the redundant UDL is removed.
+The Rust scaffolding can also be generated by the macros, meaning there are no generated `.rs`
+files involved. Types can be shared among any of the ffi modules defined in the project -
+for example, [this diplomat ffi module](https://github.com/unicode-org/icu4x/blob/7d9f89fcd7df4567e17ddd8c46810b0db287436a/ffi/diplomat/src/pluralrules.rs#L50-L51)
+uses types from a [different ffi module](https://github.com/unicode-org/icu4x/blob/7d9f89fcd7df4567e17ddd8c46810b0db287436a/ffi/diplomat/src/locale.rs#L19).
+
+It also offers better control over the stability of the API, because where the FFI is defined
+is constrained. This is [an explicit design decision](https://github.com/rust-diplomat/diplomat/blob/main/docs/design_doc.md#requirements) of diplomat.
+
+While the process for defining the type universe is different, the actual in-memory
+representation of that type universe isn't radically different from UniFFI - for example,
+here's the [definition of a Rust struct](https://github.com/rust-diplomat/diplomat/blob/main/core/src/ast/structs.rs),
+and while it is built from a `syn` struct, the final representation is independent of `syn`
+and its ast representation.
+
+## UniFFI's experience with the macro approach.
+
+Ryan tried this same macro approach for UniFFI in [#416](https://github.com/mozilla/uniffi-rs/pull/416) -
+but we struck **what's bad about this** - at least for UniFFI's use-cases: the context in which the
```

Comment:
nit: something's missing here?

Reply:
Thanks! But I can't see what's missing in this line - which probably means I should reword it :) Can you please let me know what is unclear?

Reply:
IIUC you want this sentence to treat the `**what's bad about this**` as a kind of inline link to suggest "but we struck the bad things noted here". It might be clearer as an actual link :-)

Reply:
(That is to say, I think this would read more clearly if you make the `**what's bad about this**` a kind of sub-heading in the same style as the UniFFI section above, and reference it more explicitly in the "UniFFI's experience" part).

Reply:
ah! i see